File: project/tableturf/src/starter_deck_learning.cpp
#include <iostream>
#include <algorithm>
#include <random>
#include <stdlib.h>
#include <time.h>
#include <fstream>
#include <thread>
#include <mutex>
#include <torch/script.h>
#if __has_include(<torch/mps.h>)
#include <torch/mps.h>
#endif
#include "load_model.hpp"
#include "selfplay.hpp"
#include "testplay.hpp"
#include "buffer.hpp"
#include "timer.hpp"
#include "deck_database.hpp"
template<class stage> void run(
const bool USE_TENSORRT,
int num_threads_each_gpu,
int gpu_id,
int num_games_in_parallel,int num_games_in_selfplay,int num_games_in_testplay,
int PV_ISMCTS_num_simulations,int simple_ISMCTS_num_simulations,
float diff_bonus,
float dirichlet_alpha,float eps,
char* model_path,Buffer<stage> &buffer,std::ofstream &log_out,
c10::Device device,c10::ScalarType dtype){
  Timer timer(device,gpu_id);
  static std::mutex o_mutex;//出力用排他制御
  int max_batch_size = num_games_in_parallel*4;//P1とP2、それぞれDUCT
  torch::jit::Module model = 
  (USE_TENSORRT ? 
  load_TensorRT_model<stage>(model_path,gpu_id,device,dtype,max_batch_size,max_batch_size):
  load_model(model_path,gpu_id,device,dtype)
  );
  o_mutex.lock();
  std::cout << gpu_id << ":model loaded" << std::endl;
  log_out << gpu_id << ":compilation time: " << timer.get_time() << std::endl;
  o_mutex.unlock();
  //testplay(1つのgpuだけに任せる)
  if(gpu_id == 0){
    int num_testplay_func = num_games_in_testplay/num_games_in_parallel;
    assert(num_games_in_testplay%num_games_in_parallel == 0);
    float sum_win_rate = 0;
    for(int i=0;i<num_testplay_func;i++){
      sum_win_rate += testplay<stage>(num_games_in_testplay,num_threads_each_gpu,model,device,dtype,PV_ISMCTS_num_simulations,simple_ISMCTS_num_simulations,0,std::vector<Deck>(num_games_in_testplay),std::vector<Deck>(num_games_in_testplay));
    }
    float win_rate = sum_win_rate/num_testplay_func;
    o_mutex.lock();
    log_out << "Testplay: " << win_rate << std::endl;
    log_out << gpu_id << ":testplay time: " << timer.get_time() << std::endl;
    o_mutex.unlock();
  }
  //selfplay
  static int num_games_in_selfplay_done = 0;
  static std::mutex selfplay_count_mutex;
  while(true){
    selfplay<stage>(num_games_in_parallel,num_threads_each_gpu,model,device,dtype,PV_ISMCTS_num_simulations,dirichlet_alpha,eps,diff_bonus,std::vector<Deck>(num_games_in_parallel),std::vector<Deck>(num_games_in_parallel),buffer);
    bool break_loop = false;
    selfplay_count_mutex.lock();
    num_games_in_selfplay_done += num_games_in_parallel;
    if(num_games_in_selfplay_done >= num_games_in_selfplay) break_loop = true;
    selfplay_count_mutex.unlock();
    if(break_loop) break;
  }
  o_mutex.lock();
  log_out << gpu_id << ":selfplay time: " << timer.get_time() << std::endl;
  o_mutex.unlock();
}
//ここでselfplayを回す
int main(int argc,char *argv[]){
  assert(argc == 16);
  int num_cpus = atoi(argv[1]);
  int num_gpus = atoi(argv[2]);
  char* device_str = argv[3];
  int num_games_in_parallel = atoi(argv[4]);
  int num_games_in_selfplay = atoi(argv[5]);
  int num_games_in_testplay = atoi(argv[6]);
  int buffer_size = atoi(argv[7]);
  int PV_ISMCTS_num_simulations = atoi(argv[8]);
  int simple_ISMCTS_num_simulations = atoi(argv[9]);
  float diff_bonus = atof(argv[10]);
  float dirichlet_alpha = atof(argv[11]);
  float eps = atof(argv[12]);
  char* model_path = argv[13];
  char* data_path = argv[14];
  char* log_path = argv[15];

  using stage = Main_Street;
  int num_threads_each_gpu = num_cpus*2/num_gpus;
  bool use_cuda = (strcmp(device_str,"cuda")==0);
  std::ofstream log_out(log_path,std::ios::app);//上書き
  Buffer<stage> buffer(buffer_size);//学習用データ
  // load model
  constexpr bool USE_TENSORRT = false;
  #if __has_include(<torch_tensorrt/torch_tensorrt.h>)
  if(USE_TENSORRT) std::cout << "torch_tensorrt available" << std::endl;
  #endif
  std::vector<std::thread> runner_threads;
  for(int gpu_id=0;gpu_id<num_gpus;gpu_id++){
    c10::Device device = torch::Device((use_cuda ? torch::kCUDA:torch::kMPS),gpu_id);
    c10::ScalarType dtype = (use_cuda ? torch::kFloat16:torch::kFloat32);
    runner_threads.emplace_back
    (run<stage>,
    USE_TENSORRT,
    num_threads_each_gpu,
    gpu_id,
    num_games_in_parallel,num_games_in_selfplay,num_games_in_testplay,
    PV_ISMCTS_num_simulations,simple_ISMCTS_num_simulations,
    diff_bonus,
    dirichlet_alpha,eps,
    model_path,std::ref(buffer),std::ref(log_out),
    device,dtype);
  }
  for(auto &thread:runner_threads) thread.join();
  buffer.write(data_path,false);
}

File: project/tableturf/src/starter_deck_learning.py
import torch
from torch import nn,optim
import time
import subprocess
import sys
import math
import pickle
from tqdm import tqdm
from network import network
from network.common import*
from network import datasets
from save_and_load import*

from torch.nn.parallel import DistributedDataParallel as DDP
import torch.distributed as dist
import torch.multiprocessing as mp
from torch.utils.data import DataLoader
from torch.utils.data.distributed import DistributedSampler

from config import*

def train_mps(load_model_path:str,save_model_path:str,learning_rate):
  model = load_model(load_model_path)
  # define loss function and optimizer
  loss_fn_policy_action = nn.CrossEntropyLoss()
  loss_fn_policy_redraw = nn.CrossEntropyLoss()
  loss_fn_value = nn.MSELoss()
  optimizer = optim.SGD(model.parameters(),lr=learning_rate,weight_decay=1e-4)
  
  # create dataloader
  with open(dataset_path,"rb") as file:
    train_data = pickle.load(file)
  dataloader = DataLoader(train_data,batch_size=learning_batch_size,shuffle=True,pin_memory=True,num_workers=2)

  for _ in range(n_epochs):
    with tqdm(enumerate(dataloader),total=len(dataloader)) as pbar_loss:
      for i,(data,label_policy_action,label_policy_redraw,label_value) in pbar_loss:
        data = data.to(device)
        label_policy_action = label_policy_action.to(device)
        label_policy_redraw = label_policy_redraw.to(device)
        label_value = label_value.to(device)
        # initialize grad
        optimizer.zero_grad()
        # forward
        output_policy_action,output_policy_redraw,output_value = model(data)
        # compute loss
        loss_policy_action = loss_fn_policy_action(output_policy_action,label_policy_action)
        loss_policy_redraw = loss_fn_policy_redraw(output_policy_redraw,label_policy_redraw)
        loss_value = loss_fn_value(output_value,label_value)
        # backward
        (loss_policy_action+loss_policy_redraw+loss_value).backward()
        # update parameters
        optimizer.step()
        if i==0:
          print(f"loss_policy_action:{loss_policy_action:.3f}")
          print(f"loss_policy_redraw:{loss_policy_redraw:.3f}")
          print(f"loss_value:{loss_value:.3f}")
  
  save_model(model,save_model_path)
  

def train_cuda(load_model_path:str,save_model_path:str,gpu_id:int,learning_rate:float):
  # load model
  model = load_model(load_model_path).to(gpu_id)
  # define loss function and optimizer
  loss_fn_policy_action = nn.CrossEntropyLoss()
  loss_fn_policy_redraw = nn.CrossEntropyLoss()
  loss_fn_value = nn.MSELoss()
  optimizer = optim.SGD(model.parameters(),lr=learning_rate,weight_decay=1e-4)
  # create dataloader
  with open(dataset_path,"rb") as file:
    train_data = pickle.load(file)
  num_cpus_per_gpu = num_cpus//num_gpus
  dataloader = DataLoader(train_data,batch_size=learning_batch_size,shuffle=True,num_workers=num_cpus_per_gpu,pin_memory=True)

  scaler = torch.cuda.amp.GradScaler()
  for _ in range(n_epochs):
    with tqdm(enumerate(dataloader),total=len(dataloader)) as pbar_loss:
      for i,(data,label_policy_action,label_policy_redraw,label_value) in pbar_loss:
        data = data.to(gpu_id)
        label_policy_action = label_policy_action.to(gpu_id)
        label_policy_redraw = label_policy_redraw.to(gpu_id)
        label_value = label_value.to(gpu_id)
        # initialize grad
        optimizer.zero_grad()
        with torch.autocast(device_type=device):
          # forward
          output_policy_action,output_policy_redraw,output_value = model(data)
          # compute loss
          loss_policy_action = loss_fn_policy_action(output_policy_action,label_policy_action)
          loss_policy_redraw = loss_fn_policy_redraw(output_policy_redraw,label_policy_redraw)
          loss_value = loss_fn_value(output_value,label_value)
        # backward
        scaler.scale(loss_policy_action+loss_policy_redraw+loss_value).backward()
        # update parameters
        scaler.step(optimizer)
        # update the scaler
        scaler.update()
        if i==0:
          print(f"loss_policy_action:{loss_policy_action:.3f}")
          print(f"loss_policy_redraw:{loss_policy_redraw:.3f}")
          print(f"loss_value:{loss_value:.3f}")
  
  # update current model
  save_model(model,save_model_path)

# DistributedDataParallel
def train_DDP(rank,load_model_path:str,save_model_path:str,learning_rate):
  # create default process group
  dist.init_process_group("nccl",rank=rank,world_size=num_gpus)
  # load model
  model = load_model(load_model_path).to(rank)
  # construct DDP model
  model_DDP = DDP(model,device_ids=[rank])
  # define loss function and optimizer
  loss_fn_policy_action = nn.CrossEntropyLoss()
  loss_fn_policy_redraw = nn.CrossEntropyLoss()
  loss_fn_value = nn.MSELoss()
  optimizer = optim.SGD(model_DDP.parameters(),lr=learning_rate,weight_decay=1e-4)
  # create dataloader
  with open(dataset_path,"rb") as file:
    train_data = pickle.load(file)
  sampler = DistributedSampler(dataset=train_data,num_replicas=num_gpus,rank=rank,shuffle=True)
  dataloader = DataLoader(train_data,batch_size=learning_batch_size,sampler=sampler,num_workers=1,pin_memory=True)

  scaler = torch.cuda.amp.GradScaler()
  for epoch_id in range(n_epochs):
    sampler.set_epoch(epoch_id)
    with tqdm(enumerate(dataloader),total=len(dataloader),disable=(rank!=0)) as pbar_loss:
      for i,(data,label_policy_action,label_policy_redraw,label_value) in pbar_loss:
        data = data.to(rank)
        label_policy_action = label_policy_action.to(rank)
        label_policy_redraw = label_policy_redraw.to(rank)
        label_value = label_value.to(rank)
        # initialize grad
        optimizer.zero_grad()
        with torch.autocast(device_type=device):
          # forward
          output_policy_action,output_policy_redraw,output_value = model_DDP(data)
          # compute loss
          loss_policy_action = loss_fn_policy_action(output_policy_action,label_policy_action)
          loss_policy_redraw = loss_fn_policy_redraw(output_policy_redraw,label_policy_redraw)
          loss_value = loss_fn_value(output_value,label_value)
        # backward
        scaler.scale(loss_policy_action+loss_policy_redraw+loss_value).backward()
        # update parameters
        scaler.step(optimizer)
        # update the scaler
        scaler.update()
        if rank==0 and i==0:
          print(f"loss_policy_action:{loss_policy_action:.3f}")
          print(f"loss_policy_redraw:{loss_policy_redraw:.3f}")
          print(f"loss_value:{loss_value:.3f}")
  
  dist.barrier()
  # update current model
  if rank == 0:
    # model_DDP.moduleとしないとバグるらしい
    save_model(model_DDP.module,save_model_path)

def main(current_model_path:str):
  create_new_files(current_model_path)
  with open(dataset_path,"rb") as file:
    train_data = pickle.load(file)
  with open(step_count_path,"rb") as file:
    step_count = pickle.load(file)

  # training loop
  for i in range(num_games_overall//num_games_in_selfplay):
    # C++用のモデル(policyとvalueが一体化)を作成
    create_model_cpp(current_model_path)

    # testplay & selfplay を開始
    num_cpus_for_selfplay = num_cpus*(num_gpus-1)//num_gpus
    num_gpus_for_selfplay = num_gpus-1
    command = f"{starter_deck_selfplay_program} {num_cpus_for_selfplay} {num_gpus_for_selfplay} {device} {num_games_in_parallel} {num_games_in_selfplay} {num_games_in_testplay} {buffer_size} {PV_ISMCTS_num_simulations} {simple_ISMCTS_num_simulations} {diff_bonus} {dirichlet_alpha} {eps} {model_cpp_path} {data_path} {log_path}"
    print(command)
    proc = subprocess.Popen(command,shell=True)

    # その間に1つのGPUで学習
    print("dataset size:",len(train_data))
    print("step count:",step_count)
    learning_rate = get_learning_rate(step_count)

    # update network
    if len(train_data) == 0:
      print("train_data is empty, so skip learning")
    else:
      start_learn_time = time.time()
      save_model_path = generate_model_path()
      if device=="cuda":
        # mp.spawn(train_DDP,args=(current_model_path,save_model_path,learning_rate),nprocs=num_gpus,join=True)
        learning_gpu_id = num_gpus-1
        train_cuda(current_model_path,save_model_path,learning_gpu_id,learning_rate)
        torch.cuda.synchronize()
      else:
        train_mps(current_model_path,save_model_path,learning_rate)
        torch.mps.synchronize()
      end_learn_time = time.time()

      current_model_path = save_model_path
      step_count += n_epochs*math.ceil(len(train_data)/learning_batch_size)
      with open(step_count_path,"wb") as file:
        pickle.dump(step_count,file)
      with open(log_path,"a") as log_f:
        print(f"learn time: {end_learn_time-start_learn_time:.3f}",file=log_f)
    
    # selfplay が終了
    proc.wait()
    assert proc.returncode == 0
    print("selfplay done")

    # train_dataに selfplay で得られたデータを追加
    train_data.add(data_path,buffer_size)
    with open(dataset_path,"wb") as file:
      pickle.dump(train_data,file)

  
if __name__ == "__main__":
  args = sys.argv
  current_model_path = generate_model_path()
  if len(args) >= 2:
    current_model_path = args[1]
  main(current_model_path)